uqcsbot/utils/uq_course_utils.py

```diff
@@ -16,12 +16,87 @@
 OFFERING_PARAMETER = "offer"
 
 
+class Offering:
+    """
+    A semester, campus and mode (e.g. Internal) that many courses occur within
+    """
+
+    CampusType = Literal["St Lucia", "Gatton", "Herston"]
+    # The codes used internally within UQ systems
+    campus_codes: Dict[CampusType, str] = {
+        "St Lucia": "STLUC",
+        "Gatton": "GATTN",
+        "Herston": "HERST",
+    }
+
+    ModeType = Literal["Internal", "External", "Flexible Delivery"]
+    # The codes used internally within UQ systems
+    mode_codes: Dict[ModeType, str] = {
+        "Internal": "IN",
+        "External": "EX",
+        "Flexible Delivery": "FD",
+    }
+
+    SemesterType = Literal["1", "2", "Summer"]
+    semester_codes: Dict[SemesterType, int] = {"1": 1, "2": 2, "Summer": 3}
+
+    semester: SemesterType
+    campus: CampusType
+    mode: ModeType
+
+    def __init__(
+        self,
+        semester: Optional[SemesterType],
+        campus: CampusType = "St Lucia",
+        mode: ModeType = "Internal",
+    ):
+        """
+        semester defaults to the current semester if None
+        """
+        if semester is not None:
+            self.semester = semester
+        else:
+            self.semester = "1" if datetime.today().month <= 6 else "2"
+        self.semester
+        self.campus = campus
+        self.mode = mode
+
+    def get_semester_code(self) -> int:
+        """
+        Returns the code used interally within UQ for the semester of the offering.
+        """
+        return self.semester_codes[self.semester]
+
+    def get_campus_code(self) -> str:
+        """
+        Returns the code used interally within UQ for the campus of the offering.
+        """
+        self.campus
+        return self.campus_codes[self.campus]
+
+    def get_mode_code(self) -> str:
+        """
+        Returns the code used interally within UQ for the mode of the offering.
+        """
+        return self.mode_codes[self.mode]
+
+    def get_offering_code(self) -> str:
+        """
+        Returns the hex encoded offering string (containing all offering information) for the offering.
+        """
+        return hexlify(
```

Comment:
Can we drop the `binascii` module here? I think by just .decoding it we're basically replicating `bytes.hex()` (see `text.py` for the /hexify command)